### PR TITLE
Move contract deployment

### LINKED ISSUE #394: Move contract deployment/update code into this repository
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we can't reuse deployment/update code in various projects. It was added into node in https://github.com/nspcc-dev/neofs-node/issues/2195, but you can't deploy additional contracts this way.

## Describe the solution you'd like
Port this code into a library here. Add some tests for it.

## Additional context
#384.

### PR DESCRIPTION
**Pull request by smallhive**
Closes #394

### CODE REVIEW

File: deploy/contracts.go [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+	}
+
+	// wrap the parent context into the context of the current function so that
+	// transaction wait routines do not leak
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var contractDeployer interface {
+		Sender() util.Uint160
+	}
```
Comment (AnnaShaleva) [inline deploy/contracts.go:250]: Remove it, you need a sender from it in a single place. Just store sender in this variable, that's all.

File: deploy/util.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+	blockInterval time.Duration
+
+	height atomic.Uint32
+
+	chConnLost chan struct{}
+}
+
+// newBlockchainMonitor constructs and runs monitor for the given Blockchain.
+func newBlockchainMonitor(l *zap.Logger, b Blockchain, chNewBlock chan<- struct{}) (*blockchainMonitor, error) {
```
Comment (AnnaShaleva) [inline deploy/util.go:44]: Have you completely removed this monitor from NeoFS?

File: deploy/contracts.go [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+	// Witness is not needed
+	_ Witness = iota
+	// WitnessValidators requires committee 2/3n+1.
+	WitnessValidators
+	// WitnessValidatorsAndCommittee requires WitnessValidators + committee n/2+1 with allowed NNS contract calls.
+	WitnessValidatorsAndCommittee
+)
+
+// Witness type describes witnesses for contract.
+type Witness uint8
```
Comment (roman-khimov) [inline deploy/contracts.go:39]: `WitnessSet`? `Witness` sounds too generic.

File: deploy/deploy.go [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+type ProxyContractPrm struct {
+	Common CommonDeployPrm
+}
+
+// ReputationContractPrm groups deployment parameters of the NeoFS Reputation contract.
+type ReputationContractPrm struct {
+	Common CommonDeployPrm
+}
+
+// Glagolitsa is using for alphabet contract deploy routine.
```
Comment (roman-khimov) [inline deploy/deploy.go:115]: `is used`.

File: deploy/netmap.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+	// Raw parameter value.
+	Value []byte
+}
+
+// NetworkConfiguration represents NeoFS network configuration stored
+// in the NeoFS Sidechain.
+type NetworkConfiguration struct {
+	MaxObjectSize uint64
+
```
Comment (roman-khimov) [inline deploy/netmap.go:32]: Please, don't add these spaces.

File: deploy/rand.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package deploy
+
+import (
+	crand "crypto/rand"
+	"encoding/binary"
+	mrand "math/rand"
+)
+
+var source = mrand.New(&cryptoSource{})
```
Comment (roman-khimov) [inline deploy/rand.go:9]: You certainly don't need this file at all to get a single random `uint32`.

File: deploy/contracts.go [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+	ProxyContractAddress util.Uint160
+	// SystemEmail which will be set to NNS record after contract deploy.
+	SystemEmail string
+	// DomainName for the deploying contract.
+	DomainName string
+	// NEFFile for deploying contract.
+	NEFFile nef.File
+	// ManifestFile for deploying contract (json).
+	ManifestFile manifest.Manifest
+	// DesignatedDeployer ???
```
Comment (roman-khimov) [inline deploy/contracts.go:121]: ???

File: deploy/contracts.go [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+	// pays for update transactions.
+	proxyContract util.Uint160
+	// set when syncNeoFSContractPrm relates to Proxy contract. In this case
+	// proxyContract field is unused because address is dynamically resolved within
+	// syncNeoFSContract.
+	isProxy bool
+}
+
+// ContractPrm describes parameters required to configure user's contract deployment routine.
+type ContractPrm struct {
```
Comment (roman-khimov) [inline deploy/contracts.go:99]: Doesn't look pretty, but probably OK for now.

**Review by roman-khimov — COMMENTED** [review]
I don't see `deploy_test.go` and `funds_test.go`, likely they can be kept.

File: deploy/alphabet.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	// based on blockchain
+	monitor *blockchainMonitor
+
+	committee keys.PublicKeys
+	localAcc  *wallet.Account
+}
+
+// initAlphabet designates NeoFS Alphabet role to all committee members on the
+// given Blockchain.
+func initAlphabet(ctx context.Context, prm initAlphabetPrm) error {
```
[the remark this answers is not in the document]
Comment (smallhive) [reply deploy/alphabet.go]: Renamed

File: deploy/contracts.go (hunk 3 above)
Comment (smallhive) [reply deploy/contracts.go]: Renamed

File: deploy/contracts.go (hunk 7 above)
Comment (smallhive) [reply deploy/contracts.go]: Ow, I didn't find an appropriate comment at the moment, but not sure updated one is the best one

File: deploy/contracts.go (hunk 1 above)
Comment (smallhive) [reply deploy/contracts.go:250]: Yes, one place, but different types: `*notary.Actor` and `*actor.Actor`. This _interfaced_ variable is a simple way to cover this case

File: deploy/deploy.go (hunk 4 above)
Comment (smallhive) [reply deploy/deploy.go]: Updated

File: deploy/netmap.go (hunk 5 above)
Comment (smallhive) [reply deploy/netmap.go]: Removed

File: deploy/rand.go (hunk 6 above)
Comment (smallhive) [reply deploy/rand.go]: Replaced with `rand.Uint32()`

File: deploy/util.go (hunk 2 above)
Comment (smallhive) [reply deploy/util.go:44]: Looks like yes. It wasn't used in any another place

**Comment by smallhive** [comment]
> I don't see `deploy_test.go` and `funds_test.go`, likely they can be kept.

Restored

**Review by roman-khimov — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**